src/OrleansRuntime/Core/ManagementGrain.cs

```diff
@@ -27,13 +46,17 @@ public override Task OnActivateAsync()
 
         public async Task<Dictionary<SiloAddress, SiloStatus>> GetHosts(bool onlyActive = false)
         {
-            var mTable = await GetMembershipTable();
-            var table = await mTable.ReadAll();
-            
-            var t = onlyActive ? 
-                table.Members.Where(item => item.Item1.Status == SiloStatus.Active).ToDictionary(item => item.Item1.SiloAddress, item => item.Item1.Status) :
-                table.Members.ToDictionary(item => item.Item1.SiloAddress, item => item.Item1.Status);
-            return t;
+            // If the status oracle isn't MembershipOracle, then it is assumed that it does not use IMembershipTable.
+            // In that event, return the approximate silo statuses from the status oracle.
+            if (!(this.siloStatusOracle is MembershipOracle)) return this.siloStatusOracle.GetApproximateSiloStatuses(onlyActive);
```

Comment:
Shouldn't the branching logic (or actually not really branching, but the right strategy in the concrete class) be in the oracle itself?